File: riddles-k/src/main/kotlin/com/wr/riddles/hackerearth/Logins.java
package com.wr.riddles.hackerearth;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;

public class Logins {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.valueOf(br.readLine());
        Node logins = new Node();
        for (int i = 0; i < n; i++) {
            String login = br.readLine();
            if (n == 1) {
                System.out.println(login);
                return;
            }
            Node match = add(login, logins);
            if (match == null) {
                System.out.println(login);
            } else {
                System.out.println(login + suggest(match.suggest, match));
            }
        }
    }

    static Node add(String login, Node node) {
        Node n = node;
        for (char c : login.toCharArray()) {
            n = n.nextChars.computeIfAbsent(c, k -> new Node());
        }
        if (n.isWord) {
            return n;
        } else {
            n.isWord = true;
            return null;
        }
    }

    static long suggest(int suffix, Node node) {
        int suff = suffix;
        Node existing;
        do {
            existing = add(Integer.toString(suff), node);
            if (existing != null) {
                node.suggest++;
                suff = node.suggest;
            }
        } while (existing != null);
        return suff;
    }

    static class Node {
        Map<Character, Node> nextChars = new HashMap<>();
        boolean isWord;
        int suggest;
    }

}

File: riddles-k/src/main/kotlin/com/wr/riddles/hackerearth/Logins.kt
package com.wr.riddles.hackerearth

import java.util.*

fun main(args: Array<String>) {
    val num = readLine()!!.toInt()
    if (num == 1) {
        println(readLine())
        return
    }
    val logins = Node()
    for (i in 1..num) {
        val login = readLine()!!
        print(login)
        val match = add(login, logins)
        if (match == null) {
            println()
        } else {
            println(suggest(match))
        }
    }
}

fun add(login: String, node: Node): Node? {
    var n = node
    for (c in login) {
        val next = Node()
        n = n.nextChars.putIfAbsent(c, next) ?: next
    }
    return if (n.isWord) {
        n
    } else {
        n.isWord = true
        null
    }
}

fun suggest(node: Node): Int {
    do {
        val existing = add(node.suggest.toString(), node)
        if (existing != null) {
            node.suggest++
        }
    } while (existing != null)
    return node.suggest
}

class Node {
    var nextChars: MutableMap<Char, Node> = HashMap()
    var isWord: Boolean = false
    var suggest: Int = 0
}